rfcs/react-components/styles-handbook.md

```diff
@@ -0,0 +1,771 @@
+    display: 'flex',
+    alignItems: 'center',
+    ...shorthands.padding('4px'),
+  },
+});
+// ⬇️⬇️⬇️
+// produces 6 classes (1+1+4)
+```
+
+> 💡**Tip:** To check generated classes consider to use [Try out][griffel-try-out] sandbox.
+
+However, the situation can go out of control with nested selectors, [pseudo classes][mdn-pseudo-classes]/selectors and [At-rules][mdn-at-rules] causing "CSS rules explosion":
+
+```js
+import { makeStyles, shorthands } from '@griffel/react';
+
+const useClasses = makeStyles({
+  root: {
+    ...shorthands.padding('4px'),
+    ...shorthands.margin('4px'),
+
+    ':hover': {
+      ...shorthands.padding('4px'),
+      ...shorthands.margin('4px'),
+    },
+
+    '::before': {
+      display: 'block',
+      content: "' '",
+    },
+
+    '@media (forced-colors: active)': {
+      ...shorthands.padding('4px'),
+      ...shorthands.margin('4px'),
+    },
+  },
+});
+// ⬇️⬇️⬇️
+// produces 14 classes ((4+4)+(4+4)+2+(4+4))
+```
+
+Such cases might be unavoidable by design, next section covers APIs to address this problem.
+
+## `mergeClasses()`
+
+Griffel provides [`mergeClasses() API`][griffel-merge-classes] that is expected to be used in cases with styles permutations:
+
+```js
+import { makeStyles, shorthands } from '@griffel/react';
+
+const useClasses = makeStyles({
+  root: {
+    /* styles */
+  },
+  foo: {
+    /* styles */
+  },
+  bar: {
+    /* styles */
+  },
+});
+
+function Component(props) {
+  const classes = useClasses();
+  const className = mergeClasses(
+    classes.root,
+    props.foo && classes.foo /* styles specific for "foo" */,
+    props.bar && classes.bar /* styles specific for "bar" */,
+  );
+
+  /* --- */
+}
+```
+
+### Do not concatenate classes
+
+It is not possible to simply concatenate classes returned by `useClasses()` hooks. Always use `mergeClasses()` to merge classes as results of concatenation can contain duplicated classes and lead to non-deterministic results.
+
+```js
+import { makeStyles, mergeClasses } from '@griffel/react';
+
+const useClasses = makeStyles({
+  rootA: { display: 'flex' },
+  rootB: { display: 'grid' },
+});
+
+function App(props) {
+  const classes = useClasses();
+
+  // ✅ Returns "class-display-grid"
+  const correctClasses = mergeClasses(classes.rootA, classes.rootB);
+  // 🔴 Never concatenate class strings, returns "class-display-flex class-display-grid"
+  const wrongClasses = classes.rootA + ' ' + classes.rootB;
+}
+```
+
+## `makeResetStyles`
+
+Is used to define base styles in components avoid the "CSS rules explosion" problem as generates a monolithic class:
+
+```js
+import { makeResetStyles } from '@griffel/react';
+
+const useBaseClassname = makeResetStyles({
+  padding: '4px',
+  margin: '4px',
+});
+```
+
+> ⚠️ **Note:** Only one class generated by `makeResetStyles()` can be applied to an element. Otherwise, behavior will be non-deterministic as classes merging will not be done for this case and results depend on order of insertion.
```

Comment:
```suggestion
> ⚠️ **Note:** Only one class generated by `makeResetStyles()` can be applied to an element. Otherwise, behavior will be non-deterministic since styles are not merged and deduplicated, the results will depend on order of insertion.
```